core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala

```diff
@@ -166,15 +172,15 @@ class BrokerMetadataListenerTest {
       image = newImage
     }
 
-    override def publishedOffset: Long = -1
+    override def publishedEndOffset: Long = -1
   }
 
   private val FOO_ID = Uuid.fromString("jj1G9utnTuCegi_gpnRgYw")
   private val BAR_ID = Uuid.fromString("SzN5j0LvSEaRIJHrxfMAlg")
 
   private def generateManyRecords(listener: BrokerMetadataListener,
                                   endOffset: Long): Unit = {
-    (0 to 10000).foreach { _ =>
+    (0 until 10000).foreach { _ =>
       listener.handleCommit(
         RecordTestUtils.mockBatchReader(
           endOffset,
```

Comment:
This is probably beyond the scope of this PR but it is unfortunate that this test is creating a sequence of batches that are impossible with KRaft. I assume that we are generating this many record to trigger size-based snapshots.

Reply:
Are you referring to the removing replicas changes here?

Reply:
This code generates 10k batches and sends them to the state machine but they all have the same base offset and end offset, no? You could imagine a listener implementation that ignores batches with offset that it already applied.